### DISCUSSION TITLE
React Native Google Auth

### DISCUSSION THREAD

**Discussion by BioniC187** [pocketbase/pocketbase#4131] [Q&A]
Hello,

I am unable to get google auth signin working in a react native app.

I have a brand new react-native application.
I followed the instructions mentioned here [pocketbase readme](https://github.com/pocketbase/js-sdk?tab=readme-ov-file#nodejs-via-npm)
installed pocketbase and react-native-sse and added the eventsource

I have an existing reactjs application that is working with google auth, and I am using the same pocketbase instance for this react-native app.

when I try to sign in using
`    pb.collection("users")  `
`      .authWithOAuth2({ `
`        provider: "google",    `
`        urlCallback: async url => {   `
`          console.log("urlCallbackExecuted");`
`          // open the Google sign-in url in the default browser `
`          await Linking.openURL(url);`
`        },`
`      })`
`      .then(r => {`
`        console.log("authResult", r);`
`      })`
`      .catch(e => {`
`        console.log("authError", e);`
`    });`

I get the console output 
`authError [ClientResponseError 0: Something went wrong while processing your request.]`

I have seen some people on here say they got it working in some other issues, but I am unable to get this working. Any ideas?
I have read and implemented what I can from this as well
https://github.com/pocketbase/pocketbase/discussions/2484

Any ideas? 
or does anyone have a step by step guide on how to get googleAuth working on react native?

**Comment by BioniC187** [comment]
I missed a comment where it's mentioned to log e.originalError in the catch block and now I am getting
`[Error: EventSource connect took too long.]`
as the error. However I did add in react-native-sse as mentioned and still getting the error

**Comment by BioniC187** [comment]
Okay. nevermind. 
I managed to get it figured out

**Reply by OwenLovesCoding** [reply]
can you share your code via expo snack or here, cause I am having difficulties with my code 
#reactnative

**Reply by BioniC187** [reply]
What is the error you are getting?
What have you tried for resolving the error?

**Reply by OwenLovesCoding** [reply]
I am still quite new to this so after getting my client secret for web, I tried this
 const [loading, setLoading] = useState(false);

  global.EventSource = EventSource;

  // async function handleGoogleLogin() {
  //   // const url = "http://localhost:8090/api/oauth2-redirect";
  //   try {
  //     pb.autoCancellation(false);
  //     // const authData = await pb.collection('users').authWithOAuth2({ provider: 'google' });
  //     await pb.collection("users").authWithOAuth2({
  //       provider: "google",
  //       urlCallback: (url) => {
  //         Linking.openURL(url).catch((err) => {
  //           console.log("failed to open url", err);
  //         });
  //       },
  //     });
  //     setIsvalid();
  //   } catch (e) {
  //     console.log(e);
  //   }
  // }

  // const handleGoogleLogin = async () => {
  //   setLoading(true);

  //   try {
  //     const authData = await pb.collection("users").authWithOAuth2({
  //       provider: "google",
  //       urlCallback: (url) => {
  //         console.log("hi");
  //         WebBrowser.openAuthSessionAsync(url).catch((err) => {
  //           console.log("failed to open url", err);
  //         });
  //       },
  //     });

  //     console.log(authData);
  //   } catch (error) {
  //     console.log(error);
  //     console.log(error?.originalError);
  //   }

  //   setLoading(false);
  // };

  async function handleGoogleLogin() {
    // const url = "http://localhost:8090/api/oauth2-redirect";
    try {
      pb.autoCancellation(false);
      // const authData = await pb.collection('users').authWithOAuth2({ provider: 'google' });
      await pb.collection("users").authWithOAuth2({
        provider: "google",
        urlCallback: (url) => {
          Linking.openURL(url).catch((err) => {
            console.log("failed to open url", err);
          });
        },
      });
      //setIsvalid();
    } catch (e) {
      console.log(e.originalError);
    }
  }
this is my  asyncauthstore file which I copied, import { BaseAuthStore } from "pocketbase";
import AsyncStorage from "@react-native-async-storage/async-storage";

class AsyncAuthStore extends BaseAuthStore {
  /**
   * @param {String} storageKey
   */
  constructor(storageKey = "pb_auth") {
    super();

    this.storageKey = storageKey;
    this.queue = [];

    this._enqueue(async () => {
      const raw = await AsyncStorage.getItem(this.storageKey);
      if (raw) {
        const decoded = JSON.parse(raw);
        this.save(decoded.token, decoded.model);
      }
    });
  }

  /**
   * @inheritdoc
   */
  save(token, model) {
    super.save(token, model);

    this._enqueue(async () => {
      return await AsyncStorage.setItem(
        this.storageKey,
        JSON.stringify({ token, model })
      );
    });
  }

  /**
   * @inheritdoc
   */
  clear() {
    super.clear();

    this._enqueue(async () => {
      return await AsyncStorage.removeItem(this.storageKey);
    });
  }

  /**
   * Appends an async function to the queue.
   */
  _enqueue(asyncCallback) {
    this.queue.push(asyncCallback);

    if (this.queue.length == 1) {
      this._dequeue();
    }
  }

  /**
   * Starts the queue processing.
   */
  _dequeue() {
    if (!this.queue.length) {
      return;
    }

    this.queue[0]().finally(() => {
      this.queue.shift();

      if (!this.queue.length) {
        return;
      }

      this._dequeue();
    });
  }
}

this is what it keeps giving me 
[ClientResponseError 0: Something went wrong while processing your request.]
 LOG  [TypeError: Network request failed]

**Reply by BioniC187** [reply]
Okay. So that's your code. 
What have you tried so far for resolving this error?
Are you running pocketbase on a server or locally?

**Reply by OwenLovesCoding** [reply]
I am running it locally

**Reply by BioniC187** [reply]
Alright. What about this?
>  What have you tried so far for resolving this error?

**Reply by ganigeorgiev** [reply]
_I'll allow myself to interrupt the above discussion because it seems that is going in circles._

@iraoya776 You could find a minimal snack.expo example in https://snack.expo.dev/ejfR2yMtl8xMST-zqpmhH.

The `ClientResponseError 0` and the log that comes with it indicates that your app has an issue establishing connection with the PocketBase server. This could happen for a lot of reasons but what exactly is the cause of it is difficult to pinpoint.

Your provided code sample is difficult to understand and there are not enough details where/how you are testing, what platform you are targeting, and instead of asking other users to create expo examples for you, it would've been better to paste your code in an expo example yourself so that eventually other users can assist you more easily.

With that said, couple things based on what I could understand from your comment:
- The latest JS SDK comes with builtin AsynAuthStore and you don't need to manually create it on your own. See https://github.com/pocketbase/js-sdk#asyncauthstore.

- If you are testing in emulator/phone, note that the `localhost` address you are using to initialize the JS SDK will point to the `localhost` of the mobile device, not where your PocketBase server is. To workaround this you have several options:
  - if you are testing on Android, instead of localhost you can use `10.0.2.2`
  - use the local network IP of the machine where PocketBase is running (eg. you could find it with `ifconfig`)
  - use [`ngrok`](https://ngrok.com/) to generate a public address for your PocketBase server (e.g. `ngrok http 8090`) and use that address when initializing the JS SDK (this is probably the easiest and it allow tests with real device without being part of the same network)
  
  In all cases your Google OAuth2 app callback url must be "YOUR_JS_SDK_URL/api/oauth2-redirect".

- If you are getting `EventSource connect took too long.` error then it is most likely caused by the ReactNative Flipper debug plugin (see how to disable it in https://github.com/pocketbase/pocketbase/discussions/2484#discussioncomment-6118501).

If you still have troubles making it work, please try to be more descriptive to avoid the needless back-and-forth (and eventually open a new Q&A discussion since this one is starting to get messy).

**Comment by nitinsaroha** [comment]
@BioniC187 How did you figure it out? I have the same problem. I have tried every option listen here.

**Reply by BioniC187** [reply]
What have you done so far? 

What error are you getting?
is react-native-sse setup?
Are you using in app browser or trying to launch chrome/default browser?

I can help you out, but will need some more information as to what you have done so far

**Reply by nitinsaroha** [reply]
Here's my code of the react context if it helps. I'm trying to do `github` login and it refuses to sign-in in android simulator. Release version of Android works but debug version doesn't.

I get the error: `[Error: EventSource connect took too long.]`

Environment: 
- Expo: 50
- React Native: 0.73.2


```javascript
import React from "react";
import { useStorageState } from "./useStorageState";
import { router } from "expo-router";
import PocketBase from "pocketbase";
import EventSource from "react-native-sse";
import * as WebBrowser from "expo-web-browser";
import "react-native-url-polyfill/auto";

// @ts-ignore
global.EventSource = EventSource;

const pb = new PocketBase("http://10.0.2.2:8090");
// const pb = new PocketBase("http://127.0.0.1:8090");

const AuthContext = React.createContext<{
  signIn: () => Promise<void> | null;
  signOut: () => Promise<void> | null;
  session?: string | null;
  isLoading: boolean;
}>({
  signIn: () => null,
  signOut: () => null,
  session: null,
  isLoading: false,
});

// This hook can be used to access the user info.
export function useSession() {
  const value = React.useContext(AuthContext);
  if (process.env.NODE_ENV !== "production") {
    if (!value) {
      throw new Error("useSession must be wrapped in a <SessionProvider />");
    }
  }

  return value;
}

export function SessionProvider(props: React.PropsWithChildren) {
  const [[isLoading, session], setSession] = useStorageState("session");

  const signIn = async () => {
    try {
      const authData = await pb.collection("users").authWithOAuth2({
        provider: "github",
        urlCallback: async (url) => {
          WebBrowser.openAuthSessionAsync(url)
            .then((result) => {
              console.log("Result", { result });
            })
            .catch((err) => {
              console.log("failed to open url", err);
            });
        },
      });
      setSession(authData.token);
    } catch (error) {
      console.log(JSON.stringify(error));
      // @ts-ignore
      console.log(error.originalError);
    }
  };

  return (
    <AuthContext.Provider
      value={{
        signIn,
        signOut: async () => {
          setSession(null);
          router.replace("/sign-in");
        },
        session,
        isLoading,
      }}
    >
      {props.children}
    </AuthContext.Provider>
  );
}

```

**Reply by ganigeorgiev** [reply]
@nitinsaroha For the _"EventSource connect took too long"_ error you could check the linked discussion and specifically the comments in https://github.com/pocketbase/pocketbase/discussions/2484#discussioncomment-6118501.

**Reply by BioniC187** [reply]
@nitinsaroha 
okay, if it's working in release version, do this.
go to
android/app/src/main/java/yourapp/MainApplication.kt
and comment out

`ReactNativeFlipper.initializeFlipper(this, reactNativeHost.reactInstanceManager)`

This should solve your problem.

Edit: if you still have issues, stop metro. and go a gradlew clean from inside the android folder, then start metro again.
If this doesn't work, then let me know. I can compare to what I've done and see what is missing on your app

Edit2: Just saw that you are using expo. So I don't know if you will have these files to be able to do this. Try searching around for "react native expo disable flipper" or similar
